ntp.go

```diff
@@ -365,23 +360,35 @@ func getTime(host string, opt QueryOptions) (*msg, ntpTime, error) {
 		return nil, 0, errors.New("response OriginTime != query TransmitTime") // spoofed packet?
 	}
 
+	if m.OriginTime > dst { // Go 1.9 has monotonic clock preventing that, but 1.8 has not, so it's not panic()
+		return nil, 0, errors.New("client clock tick backwards")
+	}
+
+	if m.ReceiveTime > m.TransmitTime {
+		return nil, 0, errors.New("server clock tick backwards")
+	}
+
 	return m, dst, nil
 }
 
 // TimeV returns the current time from the remote server host using the
 // requested version of the NTP protocol. On error, it returns the local time.
 // The version may be 2, 3, or 4.
 func TimeV(host string, version int) (time.Time, error) {
-	m, _, err := getTime(host, QueryOptions{Version: version})
+	m, dst, err := getTime(host, QueryOptions{Version: version})
 	if err != nil {
 		return time.Now(), err
 	}
+	r := parseTime(m, dst)
+	if !r.Validate() {
+		return time.Now(), errors.New("invalid SNTP reply")
+	}
 	// An SNTP client implementing the on-wire protocol has a single server
 	// and no dependent clients.  It can operate with any subset of the NTP
 	// on-wire protocol, the simplest approach using only the transmit
 	// timestamp of the server packet and ignoring all other fields.
 	// -- https://tools.ietf.org/html/rfc5905#section-14
-	return m.TransmitTime.Time().Local(), nil
+	return time.Now().Add(-r.ClockOffset), nil
```

Comment:
:+1:

Reply:
I hope I've used right sign for Offset :-) I tested it using gazing and I'm unsure what is correct method to express it as ntp_test.

Reply:
Yep, I've used wrong sign at first.